controlplane/unittest/segment_allocator.cpp

```diff
@@ -0,0 +1,232 @@
+	EXPECT_EQ(Block(allocator, 23), BlockStat(2, 0, 9));
+	EXPECT_EQ(Block(allocator, 27), BlockStat(1, 0, 4));
+
+	// --------------------------------------------------------------------------------------------
+	// STEP 2: - Release the length 23 segments (271, 294, 317, 340) from block 2 and then place
+	// the length 27 segments in it
+
+	EXPECT_TRUE(allocator.Free(271, 23));
+	EXPECT_EQ(Block(allocator, 23), BlockStat(1, 1, 8)); // Block 2 - in list 23
+	EXPECT_TRUE(allocator.Free(294, 23));
+	EXPECT_TRUE(allocator.Free(317, 23));
+	EXPECT_EQ(allocator.GetErrors(), no_errors);
+	EXPECT_EQ(Block(allocator, 23), BlockStat(1, 1, 6));
+
+	// release last segment of the length 23
+	EXPECT_TRUE(allocator.Free(340, 23));
+	EXPECT_EQ(Block(allocator, 23), BlockStat(1, 0, 5));
+	EXPECT_EQ(Block(allocator, 0), BlockStat(0, 1, 0)); // Block 2 - free
+
+	// start allocate segments len=27
+	EXPECT_EQ(allocator.Allocate(27), 271);
+	EXPECT_EQ(allocator.GetErrors(), no_errors);
+	EXPECT_EQ(Block(allocator, 27), BlockStat(1, 1, 5)); // Block 2 for 27
+	EXPECT_EQ(Block(allocator, 0), BlockStat(0, 0, 0));
+
+	// Allocate other segments
+	EXPECT_EQ(allocator.Allocate(27), 298);
+	EXPECT_EQ(allocator.Allocate(27), 325);
+	EXPECT_EQ(allocator.Allocate(27), 352);
+	EXPECT_EQ(allocator.Allocate(27), 0);
+	EXPECT_EQ(allocator.GetErrors(), no_errors);
+	EXPECT_EQ(Block(allocator, 27), BlockStat(2, 0, 8));
+
+	// --------------------------------------------------------------------------------------------
+	// STEP 3: - Release all segments
+
+	// Block 2
+	EXPECT_TRUE(allocator.Free(298, 27));
+	EXPECT_TRUE(allocator.Free(352, 27));
+	EXPECT_TRUE(allocator.Free(325, 27));
+	EXPECT_TRUE(allocator.Free(271, 27));
+	EXPECT_EQ(allocator.GetErrors(), no_errors);
+	EXPECT_EQ(Block(allocator, 27), BlockStat(1, 0, 4));
+	EXPECT_EQ(Block(allocator, 0), BlockStat(0, 1, 0));
+
+	// Block 0
+	EXPECT_TRUE(allocator.Free(15, 23));
+	EXPECT_TRUE(allocator.Free(38, 23));
+	EXPECT_TRUE(allocator.Free(61, 23));
+	EXPECT_TRUE(allocator.Free(84, 23));
+	EXPECT_TRUE(allocator.Free(107, 23));
+	EXPECT_EQ(allocator.GetErrors(), no_errors);
+	EXPECT_EQ(Block(allocator, 23), BlockStat(0, 0, 0));
+	EXPECT_EQ(Block(allocator, 0), BlockStat(0, 2, 0));
+
+	// Block 1
+	EXPECT_TRUE(allocator.Free(143, 27));
+	EXPECT_TRUE(allocator.Free(170, 27));
+	EXPECT_TRUE(allocator.Free(197, 27));
+	EXPECT_TRUE(allocator.Free(224, 27));
+	EXPECT_EQ(allocator.GetErrors(), no_errors);
+	EXPECT_EQ(Block(allocator, 27), BlockStat(0, 0, 0));
+	EXPECT_EQ(Block(allocator, 0), BlockStat(0, 3, 0));
+	EXPECT_EQ(allocator.GetErrors(), no_errors);
+}
+
+TEST(SegmentAllocator, BigSize)
+{
+	static constexpr uint32_t index_begin = 128;
+	static constexpr uint32_t index_end = 8 * 1024 * 1024;
+	using AllocatorBig = SegmentAllocator<index_begin, index_end, 64 * 64, 64, error_result>;
+
+	uint8_t data[sizeof(AllocatorBig)];
+	memset((void*)data, -1, sizeof(AllocatorBig));
+	AllocatorBig* allocator = new (data) AllocatorBig();
```

Comment:
I think we need `alignas(AllocatorBig)` here so the memory is properly aligned for the type being constructed.

I mean `alignas(AllocatorBig) uint8_t data[sizeof(AllocatorBig)];`

Also there's no need to cast to `(void*)`, just use `memset(data, -1, sizeof(AllocatorBig));`. 
In fact, `alignas(AllocatorBig) uint8_t data[sizeof(AllocatorBig)]  = {-1};` is enough